src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs

```diff
@@ -0,0 +1,239 @@
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Base64UrlTextEncoder = Microsoft.AspNetCore.Authentication.Base64UrlTextEncoder;
+
+namespace AspNet.Security.OAuth.Mixcloud
+{
+    public class MixcloudAuthenticationHandler : OAuthHandler<MixcloudAuthenticationOptions>
+    {
+        public MixcloudAuthenticationHandler(
+            [NotNull] IOptionsMonitor<MixcloudAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
+        {
+            var scopeParameter = properties.GetParameter<ICollection<string>>(OAuthChallengeProperties.ScopeKey);
+            var scope = scopeParameter != null ? FormatScope(scopeParameter) : FormatScope();
+
+            var parameters = new Dictionary<string, string?>
+            {
+                { "client_id", Options.ClientId },
+                { "scope", scope },
+                { "response_type", "code" }
+            };
+
+            if (Options.UsePkce)
+            {
+                var bytes = new byte[32];
+                RandomNumberGenerator.Fill(bytes);
+                var codeVerifier = Base64UrlTextEncoder.Encode(bytes);
+
+                // Store this for use during the code redemption.
+                properties.Items.Add(OAuthConstants.CodeVerifierKey, codeVerifier);
+
+                var challengeBytes = SHA256.HashData(Encoding.UTF8.GetBytes(codeVerifier));
+                var codeChallenge = WebEncoders.Base64UrlEncode(challengeBytes);
+
+                parameters[OAuthConstants.CodeChallengeKey] = codeChallenge;
+                parameters[OAuthConstants.CodeChallengeMethodKey] = OAuthConstants.CodeChallengeMethodS256;
+            }
+
+            var state = Options.StateDataFormat.Protect(properties);
+            parameters["state"] = state;
+            parameters["redirect_uri"] = QueryHelpers.AddQueryString(redirectUri, "state", state);
+
+            return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, parameters);
+        }
+
+        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
```

Comment:
All this code rings a bell, but what's the main reason with the Mixcloud provider the override is needed (compared to the default implementation in the base class)?

Reply:
This is for the bundling/unbundling of `state` as Mixcloud does not appear to support this.  I have added comments in the code to reflect this.